### PR TITLE
Add a description which port shoud be reverse proxied

### LINKED ISSUE #33559: Add an example nginx reverse proxy configuration
### Description

We have some documentations that are related to reverse proxy such as https://www.keycloak.org/server/reverseproxy and https://www.keycloak.org/server/hostname . They explain what configurations we have and some use cases but they don't have any reverse proxy configuration examples.

If we have reverse proxy configuration examples for nginx, Apache or something, users can be configured easily.

### Discussion

https://github.com/keycloak/keycloak/discussions/33441

### Motivation

I misunderstood reverse proxy configuration only with the current documentations.
If we have a reverse proxy configuration example, other uses will not misunderstand like me.

### Details

_No response_

### PR DESCRIPTION
**Pull request by kou**
I misunderstood how to configure reverse proxy with different hostname/hostname-admin. So this description will help other users.

Closes #33559

Signed-off-by: Sutou Kouhei <kou@clear-code.com>

### CODE REVIEW

**Comment by ahus1** [comment]
Thank you for this documentation example. To be able to merge this, we'll need to have a new (enhancement) issue in GitHub, as the other two you referenced are a discussion and are a closed issued moved to a discussion. When opening the new issue, you can reference those two for completeness. 

Once you created the new issue, update this PR with the new issue number (no need to create a new PR). 

I see that your example references `/js` in the reverse proxy configuration. The documentation above doesn't list it, and a short look at my browser logs doesn't show this path. Can you please double-check? 

While you configuration would work, I normally recommend to re-encrypt the traffic between nginx and Keycloak using TLS as a general rule. Your setup might be special as both services are running on the same host. Do you think you can update the example to be more general to use TLS between nginx and Keycloak? Alternatively, at the mimimum,add a reason explain that this setup uses localhost and no encryption, and that Keycloak only listens on the local address?

**Comment by kou** [comment]
Oh, sorry. I referred https://www.keycloak.org/server/reverseproxy not local file when I wrote this. I've removed `/js/`.

I've opened #33559 and referred it in the PR description and commit message.

I've removed `http-enabled true` and use `https://internal.keycloak.org:8443` not `http://127.0.0.1:8080` as the Keycloak URL. But I can't confirm that this configuration works because I don't have a Keycloak instance with HTTPS. Sorry.

**Comment by ahus1** [comment]
> But I can't confirm that this configuration works because I don't have a Keycloak instance with HTTPS.

The docs should only contain a verified configuration. If you can't verify, then it can't be part of the docs. Then you might try the other way of describing why you chose the setup that worked for you, and share all context with those reading it.

**Comment by ahus1** [comment]
> I misunderstood how to configure reverse proxy with different hostname/hostname-admin.

Intead of providing the example, there might be the path of updating the docs. If you could describe the misunderstanding that prevented you from having a working configuration, please share it here so we can discuss how the docs can be improved.

**Comment by kou** [comment]
OK. Here is my scenario:

https://www.keycloak.org/server/hostname#_exposing_the_administration_console_on_a_separate_hostname says that we can use different host names for normal operations and admin operations. I didn't think that Keycloak supports virtual host. (I didn't think that Keycloak can serve `my.keycloak.org` and `admin.my.keycloak.org` by one endpoint.) So I thought that I need to use different endpoints.

In the Keycloak log, I see Keycloak provides 2 endpoints: `0.0.0.0:8080` (because I used `http-enabled true`) and `0.0.0.0:9000`
So I thought that I need to reverse proxy both of them as `my.keycloak.org` and `admin.my.keycloak.org`. I didn't think that I need to use only one endpoint (`0.0.0.0:8080`) for both of `my.keycloak.org` and `admin.my.keycloak.org`.

What do you think about this scenario?

**Comment by domsew** [comment]
My suggestion is to create a complete example (with appropriate keycloak configs) in the form of [compose](https://github.com/compose-spec/compose-spec/tree/main) project and include it in the [quickstarts](https://github.com/keycloak/keycloak-quickstarts) repository. That way we provide reproducible example, that everyone could play with. Docs could link to this example. If you need any help with compose, just let me know.

**Comment by kou** [comment]
Thanks for your suggestion.

If we prepare TLS enabled reproducible example, we need certifications for it. How to prepare it? Should we use self-signed certifications? If we use Let's Encrypt, we need DNS records.

**Comment by domsew** [comment]
As I understood, the main goal is to demonstrate how to configure nginx as a reverse proxy, to work with Keycloak. Because this is only an example, the SSL configuration part could be skipped. In fact SSL is another topic. In `README` readers could be warned that on production they should configure HTTPS.
But if you want to add SSL to your example it can be done for many ways. The easiest seems to be generate self-signed certs with `openssl` (included in ngxinx image) or `keytool` (included in keycloak image).
As DNS you could use well known [*.localtest.me](https://readme.localtest.me/) record.

**Comment by kou** [comment]
It seems that @ahus1 wants a production ready (full TLS-ed) verified example not no-TLS example.
Let's wait for @ahus1's thought.

File: docs/guides/server/reverseproxy.adoc [hunk 1]
```diff
@@ -126,6 +126,63 @@ The following table shows the recommended paths to expose.
...
+  proxy_set_header Host $host;
+  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+  proxy_set_header X-Forwarded-Host $host;
+  proxy_set_header X-Forwarded-Proto $scheme;
+
+  # You need to add TLS configurations here
+
+  location /realms/ {
+    proxy_pass https://internal.keycloak.org:8443/realms/;
+  }
```
Comment (ahus1) [inline docs/guides/server/reverseproxy.adoc:146]:
```suggestion
  location /realms/ {
    proxy_pass https://internal.keycloak.org:8443/realms/;
     # You need to add TLS configurations here and for each additional proxy_pass 
  }
```

File: docs/guides/server/reverseproxy.adoc [hunk 2]
```diff
@@ -126,6 +126,63 @@ The following table shows the recommended paths to expose.
...
+  proxy_set_header X-Forwarded-Proto $scheme;
+
+  # You need to add TLS configurations here
+
+  # You can restrict accessable clients
+  # allow 192.168.0.0/24;
+  # deny all;
+
+  location /admin/ {
+    proxy_pass https://internal.keycloak.org:8443/admin/;
```
Comment (ahus1) [inline docs/guides/server/reverseproxy.adoc:172]:
```suggestion
    proxy_pass https://internal.keycloak.org:8443/admin/;
    # You need to add TLS configurations here and for each additional proxy_pass 
```

File: docs/guides/server/reverseproxy.adoc [hunk 3]
```diff
@@ -126,6 +126,63 @@ The following table shows the recommended paths to expose.
...
+  }
+  location /resources/ {
+    proxy_pass https://internal.keycloak.org:8443/resources/;
+  }
+}
+----
+
+In this example, we assume that {project_name} is started like the following:
+
+<@kc.start parameters="--hostname https://my.keycloak.org --hostname-admin https://admin.keycloak.org --proxy-headers xforwarded"/>
```
Comment (ahus1) [inline docs/guides/server/reverseproxy.adoc:182]:
```suggestion
<@kc.start parameters="--hostname https://my.keycloak.org --hostname-admin https://admin.keycloak.org --proxy-headers xforwarded ..."/>
```

**Review by ahus1 — COMMENTED** [review]
@kou - thank you for raising this question. My original concern was to have a configuration that doesn't use TLS for the backend that people would just copy to their own setup, change it a remote URL and then end up with an insecure setup. 

Given the discussion above, I see that adding TLS will make the example a lot larger, and and it would require additional testing. I also gave one alternative above to take the original setup, and describe what other considerations where made, although I now realize this wouldn't resolve the copy-and-paste problem. 

See below for some minor changes, and then I think this can be IMHO merged.

Having a ready-to-go compose file in the quickstarts would be a separate enhancement in a separate issue that I would still welcome. For that setup, it would be nice to have some hardcoded TLS certificates or even better a script that generates them. Maybe @domsew wants to pick this up. If we could come with a good piece of documentation when not using encryption for the backend would be ok, then we could have a second example (as long as we can expect people to not copy-and-paste it without understanding the implications).

**Comment by kou** [comment]
@ahus1 Thanks for sharing your thought. It makes sense.

I've applied your changes, adjust styles and rebase on main. Could you take a look at this again?

**Review by vmuzikar — DISMISSED** [review]
@kou Thank you for the contribution!

I agree an example how to secure admin hostname when using a reverse proxy would be a great enhancement. However, I believe this is not something that belongs directly to the reverse proxy guide. My reasoning is as follows:
* The server guides cover configuration aspects of Keycloak. Usually they are not meant to provide instructions for integration with specific 3rd party systems like nginx is.
* The example is a bit complex.
  * It's not easy to consume/try out by the user if it's just code snippets in the docs that require additional steps (like spinning up nginx).
  * It's not easily maintainable. If something in the example breaks, we don't have an automated way to know this.

Hence, I'd like to propose the following:
* Scope down this PR to contain only a generic idea of securing admin hostname applicable to any reverse proxy without the concrete example for nginx.
* Create a new ready-to-use quickstart incl. integration tests for it. (Partially related to #30828.)

**Comment by jonkoops** [comment]
> Scope down this PR to contain only a generic idea of securing admin hostname applicable to any reverse proxy without the concrete example for nginx.

I think a guide specifically for NGNIX might still be valuable, it is the source of **many** bug reports.

**Comment by jonkoops** [comment]
> Then a dedicated guide would be IMHO more suitable for visibility.

💯 this is the right place for this

**Comment by vmuzikar** [comment]
> 💯 this is the right place for this

You mean the existing proxy guide, or a dedicated new one? 🙂

**Comment by jonkoops** [comment]
A dedicated new one. The main proxy guide should link to any guides that are specific to tooling such as NGNIX, Apache or any other stack.

**Comment by kou** [comment]
Thanks for suggesting new approach.

@ahus1 What do you think about the new approach?

**Comment by ahus1** [comment]
@kou - From what I understand from the above, it acknowledges that this guide needs to be approved, and that Nginx specific advice would help users of Keycloak to set up things in a correct way.

What is now suggested goes beyond the current state of this pull request. It will eventually be more extensive, still it will take some more steps until we get there: 

1. Update the reverse proxy guide in a generic way
2. Prepare a quickstart with Nginx so we have tested templates
3. Prepare a separte guide for Nginx  

This PR would then be re-purposed to focus on (1) to improve this guide. The other items will be tacked in additional issue, either by you or other volunteers.  

Let me know if you want to give it a try for (1) and update this PR, or if I should lend a hand.

**Comment by kou** [comment]
OK. I've updated this PR for (1).

What do you think about it? If it's not correct nor doesn't include enough information. Could you lend a hand? (You can change this branch directly.)

**Comment by ahus1** [comment]
@kou - thank you for the updated documentation. I reviewed it and extended it a bit. Please have a look. If it is ok for I'll go ahead and merge it.

**Comment by kou** [comment]
Thanks. It looks good to me!
(Sorry. I'm not good at English...)

**Review by ahus1 — APPROVED** [review]

**Comment by ahus1** [comment]
@kou - no worries. I've been working with this too long so see where the documentation needs improvement, so it is good that we pair up on this.

**Comment by ahus1** [comment]
Thank you for everyone contributing to this. 

As a future step, it would be great to have an example for Nginx in our quickstarts (pining @domsew if you choose to volunteer). 

I also liked the docs from @kou where you explained which routes would be configured for which hostname. I think we miss that in the docs today. Maybe you want to volunteer and create a new issue and a PR for that, and discuss best practices for those setups.

**Pull request merged**